src/Microsoft.CodeAnalysis.Analyzers/UnitTests/MetaAnalyzers/CompareSymbolsCorrectlyTests.cs

```diff
@@ -1248,9 +1248,156 @@ public int GetHashCode(object o1, object o2)
             }.RunAsync();
         }
 
+        [Fact]
+        public async Task RS1024_CustomComparer()
+        {
+            var code = @"
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis;
+
+public class C
+{
+    public void M()
+    {
+        _ = new HashSet<ISymbol>(SymbolNameComparer.Instance);
+    }
+}
+
+internal sealed class SymbolNameComparer : EqualityComparer<ISymbol>
+{
+    private SymbolNameComparer() { }
+
+    internal static IEqualityComparer<ISymbol> Instance { get; } = new SymbolNameComparer();
+
+    public override bool Equals(ISymbol x, ISymbol y) => true;
+
+    public override int GetHashCode(ISymbol obj) => 0;
+}
+";
+
+            await new VerifyCS.Test
+            {
+                TestCode = code,
+                FixedCode = code,
+                ReferenceAssemblies = CreateNetCoreReferenceAssemblies(),
+
+            }.RunAsync();
+        }
+
+        [Fact]
+        public async Task RS1024_CustomComparer2()
+        {
+            var code = @"
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis;
+
+public class C
+{
+    public void M()
+    {
+        _ = new HashSet<ISymbol>(SymbolNameComparer.Instance);
+    }
+}
+
+internal sealed class SymbolNameComparer : EqualityComparer<ISymbol>
+{
+    private SymbolNameComparer() { }
+
+    internal static SymbolNameComparer Instance { get; } = new SymbolNameComparer();
+
+    public override bool Equals(ISymbol x, ISymbol y) => true;
+
+    public override int GetHashCode(ISymbol obj) => 0;
+}
+";
+
+            await new VerifyCS.Test
+            {
+                TestCode = code,
+                FixedCode = code,
+                ReferenceAssemblies = CreateNetCoreReferenceAssemblies(),
+
+            }.RunAsync();
+        }
+
+        [Fact]
+        public async Task RS1024_CustomComparer_ToDictionary()
+        {
+            var code = @"
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+
+public class C
+{
+    public void M(IEnumerable<ITypeSymbol> symbols)
+    {
+        _ = symbols.ToDictionary(s => s, s => s.ToDisplayString(), SymbolNameComparer.Instance);
```

Comment:
I would move this line (and the method arg) to the 1st test. We could even consider copying all the methods declared in the test `CollectionMethodsKnownToRequireComparer_DiagnosticAsync`.